### PR TITLE
Remove split of clusterRoleName and SubjectName

### PR DESCRIPTION
**Pull request by sam-nguyen7**
Fix the logic of how we create ClusterRoleBinding. 
1) Loop through every ClusterPermission
2) Create a new ClusterRoleBinding
3) Check if ClusterRoleBinding already exists on cluster
4) Update the condition if all ClusterRoleBindings were added successfully

### CODE REVIEW

File: pkg/controller/subjectpermission/subjectpermission_controller.go [hunk 1]
```diff
@@ -133,30 +132,27 @@ func (r *ReconcileSubjectPermission) Reconcile(request reconcile.Request) (recon
...
 		var CreatedClusterRoleBindingCount int
-		// loop through crClusterRoleBindingNameList
-		for _, clusterRoleBindingName := range populateCrClusterRoleBindingNameList {
-			// get the clusterRoleName by spliting the clusterRoleBindng name
-			clusterRBName := strings.Split(clusterRoleBindingName, "-")
-			clusterRoleName := clusterRBName[0]
-			subjectName := clusterRBName[1]
+		// build a clusterRoleNameList
+		var crClusterRoleNameList []string
+		for _, clusterRoleName := range instance.Spec.ClusterPermissions {
```
Comment (fahlmant) [inline pkg/controller/subjectpermission/subjectpermission_controller.go]: why are you creating a slice out of a slice? You should just range over this slice directly.

File: pkg/controller/subjectpermission/subjectpermission_controller.go [hunk 2]
```diff
@@ -133,30 +132,27 @@ func (r *ReconcileSubjectPermission) Reconcile(request reconcile.Request) (recon
...
+			newCRB := newClusterRoleBinding(a, instance.Spec.SubjectName, instance.Spec.SubjectKind)
 			err := r.client.Create(context.TODO(), newCRB)
 			if err != nil {
 				if k8serr.IsAlreadyExists(err) {
 					continue
 				}
 
 				var clusterRoleNames []string
-				clusterRoleNames = append(clusterRoleNames, clusterRoleName)
+				clusterRoleNames = append(clusterRoleNames, clusterRoleBindingName)
```
Comment (fahlmant) [inline pkg/controller/subjectpermission/subjectpermission_controller.go]: Is this var the right name? You're appending clusterRoleBinding names to a list called ClusterRole names.

File: pkg/controller/subjectpermission/subjectpermission_controller.go [hunk 3]
```diff
@@ -133,30 +132,27 @@ func (r *ReconcileSubjectPermission) Reconcile(request reconcile.Request) (recon
...
+		for _, a := range crClusterRoleNameList {
+			// add subjectName to clusterRoleName to make clusterRoleBindingName
+			clusterRoleBindingName := a + instance.Spec.SubjectName
 
 			// create a new clusterRoleBinding on cluster
-			newCRB := newClusterRoleBinding(clusterRoleName, subjectName, instance.Spec.SubjectKind)
+			newCRB := newClusterRoleBinding(a, instance.Spec.SubjectName, instance.Spec.SubjectKind)
 			err := r.client.Create(context.TODO(), newCRB)
 			if err != nil {
 				if k8serr.IsAlreadyExists(err) {
```
Comment (fahlmant) [inline pkg/controller/subjectpermission/subjectpermission_controller.go]: I would remove this and go back to checking before you run `Create` which bindings need to be created?

File: pkg/controller/subjectpermission/subjectpermission_controller.go [hunk 4]
```diff
@@ -133,30 +132,27 @@ func (r *ReconcileSubjectPermission) Reconcile(request reconcile.Request) (recon
...
-			clusterRBName := strings.Split(clusterRoleBindingName, "-")
-			clusterRoleName := clusterRBName[0]
-			subjectName := clusterRBName[1]
+		// build a clusterRoleNameList
+		var crClusterRoleNameList []string
+		for _, clusterRoleName := range instance.Spec.ClusterPermissions {
+			crClusterRoleNameList = append(crClusterRoleNameList, clusterRoleName)
+		}
+
+		for _, a := range crClusterRoleNameList {
```
[the remark this answers is not in the document]
Comment (liamwazherealso) [reply pkg/controller/subjectpermission/subjectpermission_controller.go]: Agreed

File: pkg/controller/subjectpermission/subjectpermission_controller.go [hunk 5]
```diff
@@ -132,54 +131,35 @@ func (r *ReconcileSubjectPermission) Reconcile(request reconcile.Request) (recon
...
 			}
-			// instead of updating the condition just log each successfully created ClusterRoleBinding
-			reqLogger.Info(fmt.Sprintf("Successfully created ClusterRoleBinding %s", clusterRoleBindingName))
-
-			// Add Prometheus metrics for this CR
-			localmetrics.AddPrometheusMetric(instance)
-
-			//increment counter if successfully created clusterRoleBinding
+			// if ClusterRoleBinding already exists err, add one to counter
+			clusterRoleNames = append(clusterRoleNames, clusterRoleName)
```
Comment (jharrington22) [inline pkg/controller/subjectpermission/subjectpermission_controller.go]: Does this want to be inside the `k8serr.IsAlreadyExists` block?

File: pkg/controller/subjectpermission/subjectpermission_controller.go [hunk 6]
```diff
@@ -132,54 +131,35 @@ func (r *ReconcileSubjectPermission) Reconcile(request reconcile.Request) (recon
...
-				reqLogger.Error(err, "Failed to update condition in subjectpermission controller when successfully created all cluster role bindings")
-				return reconcile.Result{}, err
-			}
-			return reconcile.Result{}, nil
+		// if ClusterRoleBinding created successfully, add one to counter
+		clusterRoleNames = append(clusterRoleNames, clusterRoleName)
+		CreatedClusterRoleBindingCount++
+	}
+	// updateCondition if all ClusterRoleBindings added successfully
+	if len(instance.Spec.ClusterPermissions) == CreatedClusterRoleBindingCount {
```
Comment (jharrington22) [inline pkg/controller/subjectpermission/subjectpermission_controller.go]: What happens if the length is no equal? Do we log or have an action?

File: pkg/controller/subjectpermission/subjectpermission_controller.go [hunk 7]
```diff
@@ -132,54 +131,36 @@ func (r *ReconcileSubjectPermission) Reconcile(request reconcile.Request) (recon
...
-			instance.Status.Conditions = controllerutil.UpdateCondition(instance.Status.Conditions, "Successfully created all ClusterRoleBindings", populateCrClusterRoleBindingNameList, true, managedv1alpha1.SubjectPermissionStateCreated, managedv1alpha1.ClusterRoleBindingCreated)
-			err = r.client.Status().Update(context.TODO(), instance)
-			if err != nil {
-				reqLogger.Error(err, "Failed to update condition in subjectpermission controller when successfully created all cluster role bindings")
-				return reconcile.Result{}, err
-			}
-			return reconcile.Result{}, nil
+		// if ClusterRoleBinding created successfully, add one to counter and append
+		clusterRoleNames = append(clusterRoleNames, clusterRoleName)
+		CreatedClusterRoleBindingCount++
```
Comment (liamwazherealso) [inline pkg/controller/subjectpermission/subjectpermission_controller.go]: `clusterRoleBindingCreatedCount` imo

File: pkg/controller/subjectpermission/subjectpermission_controller.go [hunk 8]
```diff
@@ -132,54 +131,36 @@ func (r *ReconcileSubjectPermission) Reconcile(request reconcile.Request) (recon
...
-			reqLogger.Info(fmt.Sprintf("Successfully created ClusterRoleBinding %s", clusterRoleBindingName))
-
-			// Add Prometheus metrics for this CR
-			localmetrics.AddPrometheusMetric(instance)
-
-			//increment counter if successfully created clusterRoleBinding
-			CreatedClusterRoleBindingCount++
+		} else {
+			// Created ClusterRoleBinding, update status later
+			CreatedClusterRoleBinding = true
```
[the remark this answers is not in the document]
Comment (liamwazherealso) [reply pkg/controller/subjectpermission/subjectpermission_controller.go]: `clutsterRoleBindingCreated`

File: pkg/controller/subjectpermission/subjectpermission_controller.go [hunk 9]
```diff
@@ -132,54 +131,36 @@ func (r *ReconcileSubjectPermission) Reconcile(request reconcile.Request) (recon
...
-				if k8serr.IsAlreadyExists(err) {
-					continue
-				}
-
-				var clusterRoleNames []string
-				clusterRoleNames = append(clusterRoleNames, clusterRoleName)
-
-				reqLogger.Error(err, "Failed to create clusterRoleBinding")
+	// for every ClusterPermission
+	var CreatedClusterRoleBindingCount int
```
Comment (fahlmant) [inline pkg/controller/subjectpermission/subjectpermission_controller.go]: Any specific reason to capitalize these?

File: pkg/controller/subjectpermission/subjectpermission_controller.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sam-nguyen7) [reply pkg/controller/subjectpermission/subjectpermission_controller.go]: Hmm not sure if that name is that much different @liamwazherealso . What would you suggest here for var name @fahlmant? `IsCreatedClusterRoleBinding` ?

File: pkg/controller/subjectpermission/subjectpermission_controller.go [hunk 10]
```diff
@@ -132,54 +131,36 @@ func (r *ReconcileSubjectPermission) Reconcile(request reconcile.Request) (recon
...
-				reqLogger.Error(err, "Failed to update condition in subjectpermission controller when successfully created all cluster role bindings")
-				return reconcile.Result{}, err
-			}
-			return reconcile.Result{}, nil
+		// if ClusterRoleBinding created successfully, add one to counter and append
+		clusterRoleNames = append(clusterRoleNames, clusterRoleName)
+		CreatedClusterRoleBindingCount++
+	}
+	// updateCondition if all ClusterRoleBindings added successfully
+	if CreatedClusterRoleBinding && len(instance.Spec.ClusterPermissions) == CreatedClusterRoleBindingCount {
```
[the remark this answers is not in the document]
Comment (sam-nguyen7) [reply pkg/controller/subjectpermission/subjectpermission_controller.go]: the `CreatedClusterRoleBindingCount` here will count every `clusterRoleName` regardless of if it is `already created` or not. the `&&` is to make sure we only updateCondition when we need to (i.e when we actually create the clusterRoleBinding)

File: pkg/controller/subjectpermission/subjectpermission_controller.go (hunk 9 above)
Comment (sam-nguyen7) [reply pkg/controller/subjectpermission/subjectpermission_controller.go]: no reason :) not sure why I did... will change it

File: pkg/controller/subjectpermission/subjectpermission_controller.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (fahlmant) [reply pkg/controller/subjectpermission/subjectpermission_controller.go]: ah I see, the ++ is outside of the if/else

**Pull request merged**